File: python/chebdif.py
import numpy as np
from scipy.linalg import eig
from scipy.linalg import toeplitz

def chebdif(ncheb, mder):
    """
    Calculate differentiation matrices using Chebyshev collocation.
    Returns the differentiation matrices D1, D2, .. Dmder corresponding to the
    mder-th derivative of the function f, at the ncheb Chebyshev nodes in the
    interval [-1,1].
    Parameters
    ----------
    ncheb : int, polynomial order. ncheb + 1 collocation points
    mder   : int
          maximum order of the derivative, 0 < mder <= ncheb - 1
    Returns
    -------
    x  : ndarray
         (ncheb + 1) x 1 array of Chebyshev points
    DM : ndarray
         mder x ncheb x ncheb  array of differentiation matrices
    Notes
    -----
    This function returns  mder differentiation matrices corresponding to the
    1st, 2nd, ... mder-th derivates on a Chebyshev grid of ncheb points. The
    matrices are constructed by differentiating ncheb-th order Chebyshev
    interpolants.
    The mder-th derivative of the grid function f is obtained by the matrix-
    vector multiplication
    .. math::
    f^{(m)}_i = D^{(m)}_{ij}f_j
    The code implements two strategies for enhanced accuracy suggested by
    W. Don and S. Solomonoff :
    (a) the use of trigonometric  identities to avoid the computation of
    differences x(k)-x(j)
    (b) the use of the "flipping trick"  which is necessary since sin t can
    be computed to high relative precision when t is small whereas sin (pi-t)
    cannot.
    It may, in fact, be slightly better not to implement the strategies
    (a) and (b). Please consult [3] for details.
    This function is based on code by Nikola Mirkov
    http://code.google.com/p/another-chebpy
    References
    ----------
    ..[1] B. Fornberg, Generation of Finite Difference Formulas on Arbitrarily
    Spaced Grids, Mathematics of Computation 51, no. 184 (1988): 699-706.
    ..[2] J. A. C. Weidemann and S. C. Reddy, A MATLAB Differentiation Matrix
    Suite, ACM Transactions on Mathematical Software, 26, (2000) : 465-519
    ..[3] R. Baltensperger and M. R. Trummer, Spectral Differencing With A
    Twist, SIAM Journal on Scientific Computing 24, (2002) : 1465-1487
    Examples
    --------
    The derivatives of functions is obtained by multiplying the vector of
    function values by the differentiation matrix. The N-point Chebyshev
    approximation of the first two derivatives of y = f(x) can be obtained
    as
    >>> import numpy as np
    >>> import matplotlib.pyplot as plt
    >>> import dmsuite as dm
    >>> ncheb = 32; mder = 2; pi = np.pi
    >>> x, D = dm.chebdif(ncheb, mder)        # first two derivatives
    >>> D1 = D[0,:,:]                   # first derivative
    >>> D2 = D[1,:,:]                   # second derivative
    >>> y = np.sin(2 * pi * x)                      # function at Chebyshev nodes
    >>> yd = 2 * pi * np.cos(2 * pi * x)        # theoretical first derivative
    >>> ydd = - 4 * pi ** 2 * np.sin(2 * pi * x)  # theoretical second derivative
    >>> fig, axe = plt.subplots(3, 1, sharex=True)
    >>> axe[0].plot(x, y)
    >>> axe[0].set_ylabel(r'$y$')
    >>> axe[1].plot(x, yd, '-')
    >>> axe[1].plot(x, np.dot(D1, y), 'o')
    >>> axe[1].set_ylabel(r'$y^{\prime}$')
    >>> axe[2].plot(x, ydd, '-')
    >>> axe[2].plot(x, np.dot(D2, y), 'o')
    >>> axe[2].set_xlabel(r'$x$')
    >>> axe[2].set_ylabel(r'$y^{\prime\prime}$')
    >>> plt.show()
    """

    if mder >= ncheb + 1:
        raise Exception('number of nodes must be greater than mder')

    if mder <= 0:
        raise Exception('derivative order must be at least 1')

    DM = np.zeros((mder, ncheb + 1, ncheb + 1))
    # indices used for flipping trick
    nn1 = np.int(np.floor((ncheb + 1) / 2))
    nn2 = np.int(np.ceil((ncheb + 1) / 2))
    k = np.arange(ncheb+1)
    # compute theta vector
    th = k * np.pi / ncheb

    # Compute the Chebyshev points

    # obvious way
    #x = np.cos(np.pi*np.linspace(ncheb-1,0,ncheb)/(ncheb-1))
    # W&R way
    x = np.sin(np.pi*(ncheb - 2 * np.linspace(ncheb, 0, ncheb + 1))/(2 * ncheb))
    #x = x[::-1]
    

    # Assemble the differentiation matrices
    T = np.tile(th/2, (ncheb + 1, 1))
    # trigonometric identity
    DX = 2*np.sin(T.T+T)*np.sin(T.T-T)
    # flipping trick
    DX[nn1:, :] = -np.flipud(np.fliplr(DX[0:nn2, :]))
    # diagonals of D
    DX[range(ncheb + 1), range(ncheb + 1)] = 1.
    DX = DX.T
    # matrix with entries c(k)/c(j)
    C = toeplitz((-1.)**k)
    C[0, :] *= 2
    C[-1, :] *= 2
    C[:, 0] *= 0.5
    C[:, -1] *= 0.5
    

    # Z contains entries 1/(x(k)-x(j))
    Z = 1 / DX
    # with zeros on the diagonal.
    Z[range(ncheb + 1), range(ncheb + 1)] = 0.

    # initialize differentiation matrices.
    D = np.eye(ncheb + 1)

    for ell in range(mder):
        # off-diagonals
        D = (ell + 1) * Z * (C * np.tile(np.diag(D), (ncheb + 1, 1)).T - D)
        # negative sum trick
        D[range(ncheb + 1), range(ncheb + 1)] = -np.sum(D, axis=1)
        # store current D in DM
        DM[ell, :, :] = D

    return x, DM


#import matplotlib.pyplot as plt

ncheb = 6; mder = 2; pi = np.pi
x, D = chebdif(ncheb, mder)      # first two derivatives
D1 = D[0,:,:]                       # first derivative
D2 = D[1,:,:]                       # second derivative
y = np.sin(2 * pi * x)              # function at Chebyshev nodes
yd = 2 * pi * np.cos(2 * pi * x)    # theoretical first derivative
print("")
print(D1)

File: src/chebdif.rs
extern crate ndarray;
use crate::common::*;
use ndarray::*;
use std::f64::consts::PI;

/// Calculate differentiation matrices using Chebyshev collocation.
/// Returns the differentiation matrices Dm  corresponding to the
/// der-th derivative of the function f, at the ncheb Chebyshev nodes in the
/// interval [-1,1]. Adapted from [1]
///
/// [1] https://github.com/labrosse/dmsuite
#[allow(non_snake_case)]
pub fn chebdif(n: usize, der: usize) -> (Array1<f64>, Array2<f64>) {
    assert!(n > 0, "der must be greater than 0!");
    assert!(n > der, "number of nodes must be greater than der!");
    let nf = n as f64;
    let k = Array::range(0., nf + 1.0, 1.0);

    // Compute theta vector
    let th = &k * PI / nf;

    // Chebyshev nodes
    let x = PI * (nf - 2.0 * Array::linspace(0., nf, n + 1)) / (2. * nf);
    let x = -x.mapv(f64::sin);

    // Assemble differentiation matrices
    let T = tile(&(th / 2.0));
    // trigonometric identity
    let mut Dx = 2.0 * (&T.t() + &T).mapv(f64::sin) * (&T.t() - &T).mapv(f64::sin);
    // Flipping trick (increases numerical precision)
    let nn1 = ((nf + 1.0) / 2.0).floor() as usize;
    let nn2 = ((nf + 1.0) / 2.0).ceil() as usize;
    let slice = -flipud(&fliplr(&Dx.slice(s![0..nn2, ..])));
    Dx.slice_mut(s![nn1.., ..]).assign(&slice);
    // Set diag
    set_diag(&mut Dx, 1.0);
    let Dx = Dx.t();

    // C
    let mut C = toeplitz1(&k.map(|x| (-1_i32).pow(*x as u32) as f64));
    let slice = Slice::new(0, Some(n as isize + 1), n as isize); // 1 and -1
    for a in C.slice_axis_mut(Axis(0), slice) {
        *a *= 2.;
    }
    for a in C.slice_axis_mut(Axis(1), slice) {
        *a *= 0.5;
    }

    // Z
    let mut Z = 1.0 / &Dx;
    set_diag(&mut Z, 0.);

    // initialize differentiation matrices.
    let mut D = Array2::<f64>::eye(n + 1);

    for deriv in 0..der {
        // O: np.tile(np.diag(D), (ncheb + 1, 1)).
        let O = tile(&D.diag());
        D = (deriv as f64 + 1.0) * &Z * (&C * &O.t() - &D);
        let sum = -D.sum_axis(Axis(1));
        set_diag_from_arr(&mut D, &sum);
    }
    (x, -D)
}
//
// #[cfg(test)]
// mod tests {
//     use super::*;
//     use assert_approx_eq::assert_approx_eq;
//
//     #[test]
//     fn test_chebdif() {
//         // Init
//         let n = 20;
//         let tol = 1e-4f64;
//         // Run
//         let (_, d1) = chebdif(n, 1);
//         let (x, d2) = chebdif(n, 2);
//         let f = (PI * &x).mapv(f64::sin);
//         let d1f_exp = PI * (PI * &x).mapv(f64::cos);
//         let d2f_exp = PI * PI * (PI * &x).mapv(f64::sin);
//         let d1f = d1.dot(&f);
//         let d2f = d2.dot(&f);
//         // Check
//         for (af, bf) in d1f.iter().zip(d1f_exp.iter()) {
//             assert_approx_eq!(af, bf, tol); // Check correctness
//         }
//         for (af, bf) in d2f.iter().zip(d2f_exp.iter()) {
//             assert_approx_eq!(af, bf, tol); // Check correctness
//         }
//     }
//
//     #[test]
//     #[should_panic]
//     fn test_chebdif_fail() {
//         // Init
//         let n = 20;
//         // Run
//         let (_, _) = chebdif(n, n + 1); // must panic: der>n
//     }
// }
